include/xgboost/tree_model.h

```diff
@@ -342,9 +345,9 @@ class RegTree : public Model {
     nodes_[node.RightChild()].SetParent(nid, false);
     node.SetSplit(split_index, split_value,
                   default_left);
-    // mark right child as 0, to indicate fresh leaf
-    nodes_[pleft].SetLeaf(left_leaf_weight, 0);
-    nodes_[pright].SetLeaf(right_leaf_weight, 0);
+
+    nodes_[pleft].SetLeaf(left_leaf_weight, leaf_right_child);
```

Comment:
Semantically it is a bit strange to set the left child as value `leaf_right_child`.